// natural/lib/natural/normalizers/normalizer_ja.js
/*
 Copyright (c) 2012, Guillaume Marty

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights
 to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 copies of the Software, and to permit persons to whom the Software is
 furnished to do so, subject to the following conditions:

 The above copyright notice and this permission notice shall be included in
 all copies or substantial portions of the Software.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 THE SOFTWARE.
 */

'use strict'

/**
 * Normalize Japanese inputs and expose function to perform several conversions.
 *
 * Note: The space character is treated like a roman character as it usually
 *   has the same width as them in Japanese texts.
 *
 * \@todo Replace characters range from гҲ  to гүғ, гҠҖ to гҠ° and гҮ° to гҮҝ.
 * \@todo Lazy initializations of conversionTables and converters.
 * \@todo Would fixHalfwidthKana be useful?
 *
 * Descriptions of functions exposed:
 * normalizeJapanese гҖҢе…Ёи§’гҖҚиӢұеӯ—гғ»ж•°еӯ—гӮ’гҖҢеҚҠи§’гҖҚгҖҒгҖҢеҚҠи§’гҖҚиЁҳгғ»гӮ«гӮҝгӮ«гғҠгӮ’гҖҢе…Ёи§’гҖҚгҒ«еӨүжҸӣ
 * converters.fullwidthToHalfwidth.alphabet    гҖҢе…Ёи§’гҖҚиӢұеӯ—гӮ’гҖҢеҚҠи§’гҖҚгҒ«еӨүжҸӣ
 * converters.halfwidthToFullwidth.alphabet    гҖҢеҚҠи§’гҖҚиӢұеӯ—гӮ’гҖҢе…Ёи§’гҖҚгҒ«еӨүжҸӣ
 * converters.fullwidthToHalfwidth.numbers     гҖҢе…Ёи§’гҖҚж•°еӯ—гӮ’гҖҢеҚҠи§’гҖҚгҒ«еӨүжҸӣ
 * converters.halfwidthToFullwidth.numbers     гҖҢеҚҠи§’гҖҚж•°еӯ—гӮ’гҖҢе…Ёи§’гҖҚгҒ«еӨүжҸӣ гҖҢе…Ёи§’гҖҚгӮ№гғҡгғјгӮ№гӮ’гҖҢеҚҠи§’гҖҚ
 * converters.fullwidthToHalfwidth.punctuation гҖҢе…Ёи§’гҖҚиЁҳеҸ·гӮ’гҖҢеҚҠи§’гҖҚгҒ«еӨүжҸӣ гҖҢеҚҠи§’гҖҚгӮ№гғҡгғјгӮ№гӮ’гҖҢе…Ёи§’гҖҚ
 * converters.halfwidthToFullwidth.punctuation гҖҢеҚҠи§’гҖҚиЁҳеҸ·гӮ’гҖҢе…Ёи§’гҖҚгҒ«еӨүжҸӣ
 * converters.fullwidthToHalfwidth.katakana    гҖҢе…Ёи§’гӮ«гӮҝгӮ«гғҠгҖҚгӮ’гҖҢеҚҠи§’гӮ«гӮҝгӮ«гғҠгҖҚгҒ«еӨүжҸӣ
 * converters.halfwidthToFullwidth.katakana    гҖҢеҚҠи§’гӮ«гӮҝгӮ«гғҠгҖҚгӮ’гҖҢе…Ёи§’гӮ«гӮҝгӮ«гғҠгҖҚгҒ«еӨүжҸӣ
 * converters.hiraganaToKatakana               гҖҢгӮ«гӮҝгӮ«гғҠгҖҚгӮ’гҖҢгҒІгӮүгҒҢгҒӘгҖҚгҒ«еӨүжҸӣ
 * converters.katakanaToHiragana               гҖҢгҒІгӮүгҒҢгҒӘгҖҚгӮ’гҖҢгӮ«гӮҝгӮ«гғҠгҖҚгҒ«еӨүжҸӣ
 */

const flip = require('../util/utils.js').flip
const merge = require('../util/utils.js').merge
const replacer = require('../util/utils').replacer

// From http://fernweh.jp/b/mb_convert_kana_js/
const conversionTables = {
  fullwidthToHalfwidth: {
    alphabet: {
      пҪҒ: 'a',
      пҪӮ: 'b',
      пҪғ: 'c',
      пҪ„: 'd',
      пҪ…: 'e',
      пҪҶ: 'f',
      пҪҮ: 'g',
      пҪҲ: 'h',
      пҪү: 'i',
      пҪҠ: 'j',
      пҪӢ: 'k',
      пҪҢ: 'l',
      пҪҚ: 'm',
      пҪҺ: 'n',
      пҪҸ: 'o',
      пҪҗ: 'p',
      пҪ‘: 'q',
      пҪ’: 'r',
      пҪ“: 's',
      пҪ”: 't',
      пҪ•: 'u',
      пҪ–: 'v',
      пҪ—: 'w',
      пҪҳ: 'x',
      пҪҷ: 'y',
      пҪҡ: 'z',
      пјЎ: 'A',
      пјў: 'B',
      пјЈ: 'C',
      пјӨ: 'D',
      пјҘ: 'E',
      пјҰ: 'F',
      пј§: 'G',
      пјЁ: 'H',
      пј©: 'I',
      пјӘ: 'J',
      пј«: 'K',
      пј¬: 'L',
      пјӯ: 'M',
      пј®: 'N',
      пјҜ: 'O',
      пј°: 'P',
      пјұ: 'Q',
      пјІ: 'R',
      пјі: 'S',
      пјҙ: 'T',
      пјө: 'U',
      пј¶: 'V',
      пј·: 'W',
      пјё: 'X',
      пј№: 'Y',
      пјә: 'Z',
      'гҖҖ': ' ' // Fullwidth space
    },

    numbers: {
      'пјҗ': '0',
      'пј‘': '1',
      'пј’': '2',
      'пј“': '3',
      'пј”': '4',
      'пј•': '5',
      'пј–': '6',
      'пј—': '7',
      'пјҳ': '8',
      'пјҷ': '9'
    },

    symbol: {
      'пјҝ': '_',
      'пјҚ': '-',
      'пјҢ': ',',
      'пјӣ': ';',
      'пјҡ': ':',
      'пјҒ': '!',
      'пјҹ': '?',
      'пјҺ': '.',
      'пјҲ': '(',
      'пјү': ')',
      'пј»': '[',
      'пјҪ': ']',
      'пҪӣ': '{',
      'пҪқ': '}',
      'пј ': '@',
      'пјҠ': '*',
      'пјј': '\\',
      'пјҸ': '/',
      'пјҶ': '&',
      'пјғ': '#',
      'пј…': '%',
      'пҪҖ': '`',
      'пјҫ': '^',
      'пјӢ': '+',
      'пјң': '<',
      'пјқ': '=',
      'пјһ': '>',
      'пҪң': '|',
      // Never converted: 'пҪһ': '~',
      'вүӘ': 'В«',
      'вү«': 'В»',
      'в”Җ': '-',
      'пј„': '$',
      'пјӮ': '"'
    },

    purePunctuation: {
      'гҖҒ': 'пҪӨ',
      'гҖӮ': 'пҪЎ',
      'гғ»': 'пҪҘ',
      'гҖҢ': 'пҪў',
      'гҖҚ': 'пҪЈ'
    },

    punctuation: {},

    katakana: {
      гӮӣ: 'пҫһ',
      гӮң: 'пҫҹ',
      гғј: 'пҪ°',

      гғҙ: 'пҪіпҫһ',
      гӮ¬: 'пҪ¶пҫһ',
      гӮ®: 'пҪ·пҫһ',
      гӮ°: 'пҪёпҫһ',
      гӮІ: 'пҪ№пҫһ',
      гӮҙ: 'пҪәпҫһ',
      гӮ¶: 'пҪ»пҫһ',
      гӮё: 'пҪјпҫһ',
      гӮә: 'пҪҪпҫһ',
      гӮј: 'пҪҫпҫһ',
      гӮҫ: 'пҪҝпҫһ',
      гғҖ: 'пҫҖпҫһ',
      гғӮ: 'пҫҒпҫһ',
      гғ…: 'пҫӮпҫһ',
      гғҮ: 'пҫғпҫһ',
      гғү: 'пҫ„пҫһ',
      гғҗ: 'пҫҠпҫһ',
      гғ‘: 'пҫҠпҫҹ',
      гғ“: 'пҫӢпҫһ',
      гғ”: 'пҫӢпҫҹ',
      гғ–: 'пҫҢпҫһ',
      гғ—: 'пҫҢпҫҹ',
      гғҷ: 'пҫҚпҫһ',
      гғҡ: 'пҫҚпҫҹ',
      гғң: 'пҫҺпҫһ',
      гғқ: 'пҫҺпҫҹ',

      гӮЎ: 'пҪ§',
      гӮў: 'пҪұ',
      гӮЈ: 'пҪЁ',
      гӮӨ: 'пҪІ',
      гӮҘ: 'пҪ©',
      гӮҰ: 'пҪі',
      гӮ§: 'пҪӘ',
      гӮЁ: 'пҪҙ',
      гӮ©: 'пҪ«',
      гӮӘ: 'пҪө',
      гӮ«: 'пҪ¶',
      гӮӯ: 'пҪ·',
      гӮҜ: 'пҪё',
      гӮұ: 'пҪ№',
      гӮі: 'пҪә',
      гӮө: 'пҪ»',
      гӮ·: 'пҪј',
      гӮ№: 'пҪҪ',
      гӮ»: 'пҪҫ',
      гӮҪ: 'пҪҝ',
      гӮҝ: 'пҫҖ',
      гғҒ: 'пҫҒ',
      гғғ: 'пҪҜ',
      гғ„: 'пҫӮ',
      гғҶ: 'пҫғ',
      гғҲ: 'пҫ„',
      гғҠ: 'пҫ…',
      гғӢ: 'пҫҶ',
      гғҢ: 'пҫҮ',
      гғҚ: 'пҫҲ',
      гғҺ: 'пҫү',
      гғҸ: 'пҫҠ',
      гғ’: 'пҫӢ',
      гғ•: 'пҫҢ',
      гғҳ: 'пҫҚ',
      гғӣ: 'пҫҺ',
      гғһ: 'пҫҸ',
      гғҹ: 'пҫҗ',
      гғ : 'пҫ‘',
      гғЎ: 'пҫ’',
      гғў: 'пҫ“',
      гғЈ: 'пҪ¬',
      гғӨ: 'пҫ”',
      гғҘ: 'пҪӯ',
      гғҰ: 'пҫ•',
      гғ§: 'пҪ®',
      гғЁ: 'пҫ–',
      гғ©: 'пҫ—',
      гғӘ: 'пҫҳ',
      гғ«: 'пҫҷ',
      гғ¬: 'пҫҡ',
      гғӯ: 'пҫӣ',
      гғҜ: 'пҫң',
      гғІ: 'пҪҰ',
      гғі: 'пҫқ'
    }
  },

  halfwidthToFullwidth: {}
}

const fixFullwidthKana = {
  гӮқгӮӣ: 'гӮһ',
  гғҪгӮӣ: 'гғҫ',

  гҒҶгӮӣ: 'гӮ”',
  гҒӢгӮӣ: 'гҒҢ',
  гҒҚгӮӣ: 'гҒҺ',
  гҒҸгӮӣ: 'гҒҗ',
  гҒ‘гӮӣ: 'гҒ’',
  гҒ“гӮӣ: 'гҒ”',
  гҒ•гӮӣ: 'гҒ–',
  гҒ—гӮӣ: 'гҒҳ',
  гҒҷгӮӣ: 'гҒҡ',
  гҒӣгӮӣ: 'гҒң',
  гҒқгӮӣ: 'гҒһ',
  гҒҹгӮӣ: 'гҒ ',
  гҒЎгӮӣ: 'гҒў',
  гҒӨгӮӣ: 'гҒҘ',
  гҒҰгӮӣ: 'гҒ§',
  гҒЁгӮӣ: 'гҒ©',
  гҒҜгӮӣ: 'гҒ°',
  гҒҜгӮң: 'гҒұ',
  гҒІгӮӣ: 'гҒі',
  гҒІгӮң: 'гҒҙ',
  гҒөгӮӣ: 'гҒ¶',
  гҒөгӮң: 'гҒ·',
  гҒёгӮӣ: 'гҒ№',
  гҒёгӮң: 'гҒә',
  гҒ»гӮӣ: 'гҒј',
  гҒ»гӮң: 'гҒҪ',
  гҒЈгҒӘ: 'гӮ“гҒӘ',
  гҒЈгҒ«: 'гӮ“гҒ«',
  гҒЈгҒ¬: 'гӮ“гҒ¬',
  гҒЈгҒӯ: 'гӮ“гҒӯ',
  гҒЈгҒ®: 'гӮ“гҒ®',

  гӮҰгӮӣ: 'гғҙ',
  гӮ«гӮӣ: 'гӮ¬',
  гӮӯгӮӣ: 'гӮ®',
  гӮҜгӮӣ: 'гӮ°',
  гӮұгӮӣ: 'гӮІ',
  гӮігӮӣ: 'гӮҙ',
  гӮөгӮӣ: 'гӮ¶',
  гӮ·гӮӣ: 'гӮё',
  гӮ№гӮӣ: 'гӮә',
  гӮ»гӮӣ: 'гӮј',
  гӮҪгӮӣ: 'гӮҫ',
  гӮҝгӮӣ: 'гғҖ',
  гғҒгӮӣ: 'гғӮ',
  гғ„гӮӣ: 'гғ…',
  гғҶгӮӣ: 'гғҮ',
  гғҲгӮӣ: 'гғү',
  гғҸгӮӣ: 'гғҗ',
  гғҸгӮң: 'гғ‘',
  гғ’гӮӣ: 'гғ“',
  гғ’гӮң: 'гғ”',
  гғ•гӮӣ: 'гғ–',
  гғ•гӮң: 'гғ—',
  гғҳгӮӣ: 'гғҷ',
  гғҳгӮң: 'гғҡ',
  гғӣгӮӣ: 'гғң',
  гғӣгӮң: 'гғқ',
  гғғгғҠ: 'гғігғҠ',
  гғғгғӢ: 'гғігғӢ',
  гғғгғҢ: 'гғігғҢ',
  гғғгғҚ: 'гғігғҚ',
  гғғгғҺ: 'гғігғҺ'
}

const fixCompositeSymbolsTable = {
  'гӢҖ': '1жңҲ',
  'гӢҒ': '2жңҲ',
  'гӢӮ': '3жңҲ',
  'гӢғ': '4жңҲ',
  'гӢ„': '5жңҲ',
  'гӢ…': '6жңҲ',
  'гӢҶ': '7жңҲ',
  'гӢҮ': '8жңҲ',
  'гӢҲ': '9жңҲ',
  'гӢү': '10жңҲ',
  'гӢҠ': '11жңҲ',
  'гӢӢ': '12жңҲ',

  'гҸ ': '1ж—Ҙ',
  'гҸЎ': '2ж—Ҙ',
  'гҸў': '3ж—Ҙ',
  'гҸЈ': '4ж—Ҙ',
  'гҸӨ': '5ж—Ҙ',
  'гҸҘ': '6ж—Ҙ',
  'гҸҰ': '7ж—Ҙ',
  'гҸ§': '8ж—Ҙ',
  'гҸЁ': '9ж—Ҙ',
  'гҸ©': '10ж—Ҙ',
  'гҸӘ': '11ж—Ҙ',
  'гҸ«': '12ж—Ҙ',
  'гҸ¬': '13ж—Ҙ',
  'гҸӯ': '14ж—Ҙ',
  'гҸ®': '15ж—Ҙ',
  'гҸҜ': '16ж—Ҙ',
  'гҸ°': '17ж—Ҙ',
  'гҸұ': '18ж—Ҙ',
  'гҸІ': '19ж—Ҙ',
  'гҸі': '20ж—Ҙ',
  'гҸҙ': '21ж—Ҙ',
  'гҸө': '22ж—Ҙ',
  'гҸ¶': '23ж—Ҙ',
  'гҸ·': '24ж—Ҙ',
  'гҸё': '25ж—Ҙ',
  'гҸ№': '26ж—Ҙ',
  'гҸә': '27ж—Ҙ',
  'гҸ»': '28ж—Ҙ',
  'гҸј': '29ж—Ҙ',
  'гҸҪ': '30ж—Ҙ',
  'гҸҫ': '31ж—Ҙ',

  'гҚҳ': '0зӮ№',
  'гҚҷ': '1зӮ№',
  'гҚҡ': '2зӮ№',
  'гҚӣ': '3зӮ№',
  'гҚң': '4зӮ№',
  'гҚқ': '5зӮ№',
  'гҚһ': '6зӮ№',
  'гҚҹ': '7зӮ№',
  'гҚ ': '8зӮ№',
  'гҚЎ': '9зӮ№',
  'гҚў': '10зӮ№',
  'гҚЈ': '11зӮ№',
  'гҚӨ': '12зӮ№',
  'гҚҘ': '13зӮ№',
  'гҚҰ': '14зӮ№',
  'гҚ§': '15зӮ№',
  'гҚЁ': '16зӮ№',
  'гҚ©': '17зӮ№',
  'гҚӘ': '18зӮ№',
  'гҚ«': '19зӮ№',
  'гҚ¬': '20зӮ№',
  'гҚӯ': '21зӮ№',
  'гҚ®': '22зӮ№',
  'гҚҜ': '23зӮ№',
  'гҚ°': '24зӮ№',

  'гҚ»': 'е№іжҲҗ',
  'гҚј': 'жҳӯе’Ң',
  'гҚҪ': 'еӨ§жӯЈ',
  'гҚҫ': 'жҳҺжІ»',
  'гҚҝ': 'ж ӘејҸдјҡзӨҫ',

  'гҢҖ': 'гӮўгғ‘гғјгғҲ',
  'гҢҒ': 'гӮўгғ«гғ•гӮЎ',
  'гҢӮ': 'гӮўгғігғҡгӮў',
  'гҢғ': 'гӮўгғјгғ«',
  'гҢ„': 'гӮӨгғӢгғігӮ°',
  'гҢ…': 'гӮӨгғігғҒ',
  'гҢҶ': 'гӮҰгӮӘгғі',
  'гҢҮ': 'гӮЁгӮ№гӮҜгғјгғү',
  'гҢҲ': 'гӮЁгғјгӮ«гғј',
  'гҢү': 'гӮӘгғігӮ№',
  'гҢҠ': 'гӮӘгғјгғ ',
  'гҢӢ': 'гӮ«гӮӨгғӘ', // жө·йҮҢ
  'гҢҢ': 'гӮ«гғ©гғғгғҲ',
  'гҢҚ': 'гӮ«гғӯгғӘгғј',
  'гҢҺ': 'гӮ¬гғӯгғі',
  'гҢҸ': 'гӮ¬гғігғһ',
  'гҢҗ': 'гӮ®гӮ¬',
  'гҢ‘': 'гӮ®гғӢгғј',
  'гҢ’': 'гӮӯгғҘгғӘгғј',
  'гҢ“': 'гӮ®гғ«гғҖгғј',
  'гҢ”': 'гӮӯгғӯ',
  'гҢ•': 'гӮӯгғӯгӮ°гғ©гғ ',
  'гҢ–': 'гӮӯгғӯгғЎгғјгғҲгғ«',
  'гҢ—': 'гӮӯгғӯгғҜгғғгғҲ',
  'гҢҳ': 'гӮ°гғ©гғ ',
  'гҢҷ': 'гӮ°гғ©гғ гғҲгғі',
  'гҢҡ': 'гӮҜгғ«гӮјгӮӨгғӯ',
  'гҢӣ': 'гӮҜгғӯгғјгғҚ',
  'гҢң': 'гӮұгғјгӮ№',
  'гҢқ': 'гӮігғ«гғҠ',
  'гҢһ': 'гӮігғјгғқ',
  'гҢҹ': 'гӮөгӮӨгӮҜгғ«',
  'гҢ ': 'гӮөгғігғҒгғјгғ ',
  'гҢЎ': 'гӮ·гғӘгғігӮ°',
  'гҢў': 'гӮ»гғігғҒ',
  'гҢЈ': 'гӮ»гғігғҲ',
  'гҢӨ': 'гғҖгғјгӮ№',
  'гҢҘ': 'гғҮгӮ·',
  'гҢҰ': 'гғүгғ«',
  'гҢ§': 'гғҲгғі',
  'гҢЁ': 'гғҠгғҺ',
  'гҢ©': 'гғҺгғғгғҲ',
  'гҢӘ': 'гғҸгӮӨгғ„',
  'гҢ«': 'гғ‘гғјгӮ»гғігғҲ',
  'гҢ¬': 'гғ‘гғјгғ„',
  'гҢӯ': 'гғҗгғјгғ¬гғ«',
  'гҢ®': 'гғ”гӮўгӮ№гғҲгғ«',
  'гҢҜ': 'гғ”гӮҜгғ«',
  'гҢ°': 'гғ”гӮі',
  'гҢұ': 'гғ“гғ«',
  'гҢІ': 'гғ•гӮЎгғ©гғғгғү',
  'гҢі': 'гғ•гӮЈгғјгғҲ',
  'гҢҙ': 'гғ–гғғгӮ·гӮ§гғ«',
  'гҢө': 'гғ•гғ©гғі',
  'гҢ¶': 'гғҳгӮҜгӮҝгғјгғ«',
  'гҢ·': 'гғҡгӮҪ',
  'гҢё': 'гғҡгғӢгғ’',
  'гҢ№': 'гғҳгғ«гғ„',
  'гҢә': 'гғҡгғігӮ№',
  'гҢ»': 'гғҡгғјгӮё',
  'гҢј': 'гғҷгғјгӮҝ',
  'гҢҪ': 'гғқгӮӨгғігғҲ',
  'гҢҫ': 'гғңгғ«гғҲ',
  'гҢҝ': 'гғӣгғі',
  'гҚҖ': 'гғқгғігғү',
  'гҚҒ': 'гғӣгғјгғ«',
  'гҚӮ': 'гғӣгғјгғі',
  'гҚғ': 'гғһгӮӨгӮҜгғӯ',
  'гҚ„': 'гғһгӮӨгғ«',
  'гҚ…': 'гғһгғғгғҸ',
  'гҚҶ': 'гғһгғ«гӮҜ',
  'гҚҮ': 'гғһгғігӮ·гғ§гғі',
  'гҚҲ': 'гғҹгӮҜгғӯгғі',
  'гҚү': 'гғҹгғӘ',
  'гҚҠ': 'гғҹгғӘгғҗгғјгғ«',
  'гҚӢ': 'гғЎгӮ¬',
  'гҚҢ': 'гғЎгӮ¬гғҲгғі',
  'гҚҚ': 'гғЎгғјгғҲгғ«',
  'гҚҺ': 'гғӨгғјгғү',
  'гҚҸ': 'гғӨгғјгғ«',
  'гҚҗ': 'гғҰгӮўгғі',
  'гҚ‘': 'гғӘгғғгғҲгғ«',
  'гҚ’': 'гғӘгғ©',
  'гҚ“': 'гғ«гғ”гғј',
  'гҚ”': 'гғ«гғјгғ–гғ«',
  'гҚ•': 'гғ¬гғ ',
  'гҚ–': 'гғ¬гғігғҲгӮІгғі',
  'гҚ—': 'гғҜгғғгғҲ'
}

// punctuation is pure_punctuation
conversionTables.fullwidthToHalfwidth.punctuation = merge(
  conversionTables.fullwidthToHalfwidth.symbol,
  conversionTables.fullwidthToHalfwidth.purePunctuation
)

// Fill in the conversion tables with the flipped tables.
conversionTables.halfwidthToFullwidth.alphabet = flip(conversionTables.fullwidthToHalfwidth.alphabet)
conversionTables.halfwidthToFullwidth.numbers = flip(conversionTables.fullwidthToHalfwidth.numbers)
conversionTables.halfwidthToFullwidth.symbol = flip(conversionTables.fullwidthToHalfwidth.symbol)
conversionTables.halfwidthToFullwidth.purePunctuation = flip(conversionTables.fullwidthToHalfwidth.purePunctuation)
conversionTables.halfwidthToFullwidth.punctuation = flip(conversionTables.fullwidthToHalfwidth.punctuation)
conversionTables.halfwidthToFullwidth.katakana = flip(conversionTables.fullwidthToHalfwidth.katakana)

// Build the normalization table.
conversionTables.normalize = merge(
  conversionTables.fullwidthToHalfwidth.alphabet,
  conversionTables.fullwidthToHalfwidth.numbers,
  conversionTables.fullwidthToHalfwidth.symbol,
  conversionTables.halfwidthToFullwidth.purePunctuation,
  conversionTables.halfwidthToFullwidth.katakana
)

const converters = {
  fullwidthToHalfwidth: {
    alphabet: replacer(conversionTables.fullwidthToHalfwidth.alphabet),
    numbers: replacer(conversionTables.fullwidthToHalfwidth.numbers),
    symbol: replacer(conversionTables.fullwidthToHalfwidth.symbol),
    purePunctuation: replacer(conversionTables.fullwidthToHalfwidth.purePunctuation),
    punctuation: replacer(conversionTables.fullwidthToHalfwidth.punctuation),
    katakana: replacer(conversionTables.fullwidthToHalfwidth.katakana)
  },

  halfwidthToFullwidth: {
    alphabet: replacer(conversionTables.halfwidthToFullwidth.alphabet),
    numbers: replacer(conversionTables.halfwidthToFullwidth.numbers),
    symbol: replacer(conversionTables.halfwidthToFullwidth.symbol),
    purePunctuation: replacer(conversionTables.halfwidthToFullwidth.purePunctuation),
    punctuation: replacer(conversionTables.halfwidthToFullwidth.punctuation),
    katakana: replacer(conversionTables.halfwidthToFullwidth.katakana)
  },

  fixFullwidthKana: replacer(fixFullwidthKana),
  normalize: replacer(conversionTables.normalize)
}

const fixCompositeSymbols = replacer(fixCompositeSymbolsTable)

/**
 * Convert hiragana to fullwidth katakana.
 * According to http://jsperf.com/converting-japanese, these implementations are
 * faster than using lookup tables.
 *
 * @param {string} str A string.
 * @return {string} A string not containing hiragana.
 */
converters.hiraganaToKatakana = function (str) {
  str = converters.halfwidthToFullwidth.katakana(str)
  str = converters.fixFullwidthKana(str)

  str = str.replace(/гӮқ/g, 'гғҪ')
  str = str.replace(/гӮһ/g, 'гғҫ')
  // str = str.replace(/?/g, 'рӣҖҖ'); // Letter archaic E

  str = str.replace(/[гҒҒ-гӮ–]/g, function (str) {
    return String.fromCharCode(str.charCodeAt(0) + 96)
  })

  return str
}

/**
 * Convert katakana to hiragana.
 *
 * @param {string} str A string.
 * @return {string} A string not containing katakana.
 */
converters.katakanaToHiragana = function (str) {
  str = converters.halfwidthToFullwidth.katakana(str)
  str = converters.fixFullwidthKana(str)

  str = str.replace(/гғҪ/g, 'гӮқ')
  str = str.replace(/гғҫ/g, 'гӮһ')
  // str = str.replace(/?/g, 'рӣҖҒ'); // Letter archaic E

  str = str.replace(/[гӮЎ-гғ¶]/g, function (str) {
    return String.fromCharCode(str.charCodeAt(0) - 96)
  })

  return str
}

/**
 * Fix kana and apply the following processes;
 * * Replace repeat characters
 * * Alphabet to halfwidth
 * * Numbers to halfwidth
 * * Punctuation to fullwidth
 * * Katakana to fullwidth
 * * Fix fullwidth kana
 * * Replace composite symbols
 *
 * @param {string} str
 * @return {string}
 */
const normalizeJa = function (str) {
  // Replace repeat characters.
  str = str
    .replace(/(..)гҖ…гҖ…/g, '$1$1')
    .replace(/(.)гҖ…/g, '$1$1')

  str = converters.normalize(str)
  str = converters.fixFullwidthKana(str)

  // Replace composite symbols.
  str = fixCompositeSymbols(str)

  return str
}

exports.normalizeJa = normalizeJa
exports.converters = converters
