// natural/lib/natural/stemmers/porter_stemmer_ru.js
/*
Copyright (c) 2012, Polyakov Vladimir, Chris Umbel

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
THE SOFTWARE.
*/

'use strict'

const Stemmer = require('./stemmer_ru')

const PorterStemmer = new Stemmer()
module.exports = PorterStemmer

function attemptReplacePatterns (token, patterns) {
  let replacement = null
  let i = 0; let isReplaced = false
  while ((i < patterns.length) && !isReplaced) {
    if (patterns[i][0].test(token)) {
      replacement = token.replace(patterns[i][0], patterns[i][1])
      isReplaced = true
    }
    i++
  }
  return replacement
}

function perfectiveGerund (token) {
  const result = attemptReplacePatterns(token, [
    [/[邪褟]胁(褕懈|褕懈褋褜)$/g, ''],
    [/(懈胁|懈胁褕懈|懈胁褕懈褋褜|褘胁褕懈|褘胁褕懈褋褜|褘胁)$/g, '']
  ])
  return result
}

function adjectival (token) {
  let result = adjective(token)
  if (result != null) {
    const pariticipleResult = participle(result)
    result = pariticipleResult || result
  }
  return result
}

function adjective (token) {
  const result = attemptReplacePatterns(token, [
    [/(械械|懈械|褘械|芯械|懈屑懈|褘屑懈|械泄|懈泄|褘泄|芯泄|械屑|懈屑|褘屑|芯屑|械谐芯|芯谐芯|械屑褍|芯屑褍|懈褏|褘褏|褍褞|褞褞|邪褟|褟褟|芯褞|械褞)$/g, '']
  ])
  return result
}

function participle (token) {
  const result = attemptReplacePatterns(token, [
    [/([邪褟])(械屑|薪薪|胁褕|褞褖|褖)$/g, '$1'],
    [/(懈胁褕|褘胁褕|褍褞褖)$/g, '']
  ])
  return result
}

function reflexive (token) {
  const result = attemptReplacePatterns(token, [
    [/(褋褟|褋褜)$/g, '']
  ])
  return result
}

function verb (token) {
  const result = attemptReplacePatterns(token, [
    [/([邪褟])(谢邪|薪邪|械褌械|泄褌械|谢懈|泄|谢|械屑|薪|谢芯|薪芯|械褌|褞褌|薪褘|褌褜|械褕褜|薪薪芯)$/g, '$1'],
    [/(懈谢邪|褘谢邪|械薪邪|械泄褌械|褍泄褌械|懈褌械|懈谢懈|褘谢懈|械泄|褍泄|懈谢|褘谢|懈屑|褘屑|械薪|懈谢芯|褘谢芯|械薪芯|褟褌|褍械褌|懈褌|褘褌|械薪褘|懈褌褜|褘褌褜|懈褕褜|褍褞|褞)$/g, '']
  ])
  return result
}

function noun (token) {
  const result = attemptReplacePatterns(token, [
    [/(邪|械胁|芯胁|懈械|褜械|械|懈褟屑懈|褟屑懈|邪屑懈|械懈|懈懈|懈|懈械泄|械泄|芯泄|懈泄|泄|懈褟屑|褟屑|懈械屑|械屑|邪屑|芯屑|芯|褍|邪褏|懈褟褏|褟褏|褘|褜|懈褞|褜褞|褞|懈褟|褜褟|褟)$/g, '']
  ])
  return result
}

function superlative (token) {
  const result = attemptReplacePatterns(token, [
    [/(械泄褕|械泄褕械)$/g, '']
  ])
  return result
}

function derivational (token) {
  const result = attemptReplacePatterns(token, [
    [/(芯褋褌|芯褋褌褜)$/g, '']
  ])
  return result
}

// perform full stemming algorithm on a single word
PorterStemmer.stem = function (token) {
  token = token.toLowerCase().replace(/褢/g, '械')
  const volwesRegexp = /^(.*?[邪械懈芯褞褟褍褘懈褝])(.*)$/g
  let RV = volwesRegexp.exec(token)
  if (!RV || RV.length < 3) {
    return token
  }
  const head = RV[1]
  RV = RV[2]
  volwesRegexp.lastIndex = 0
  const R2 = volwesRegexp.exec(RV)
  let result = perfectiveGerund(RV)
  if (result === null) {
    const resultReflexive = reflexive(RV) || RV
    result = adjectival(resultReflexive)
    if (result === null) {
      result = verb(resultReflexive)
      if (result === null) {
        result = noun(resultReflexive)
        if (result === null) {
          result = resultReflexive
        }
      }
    }
  }
  result = result.replace(/懈$/g, '')
  let derivationalResult = result
  if (R2 && R2[2]) {
    derivationalResult = derivational(R2[2])
    if (derivationalResult != null) {
      derivationalResult = derivational(result)
    } else {
      derivationalResult = result
    }
  }

  let superlativeResult = superlative(derivationalResult) || derivationalResult

  superlativeResult = superlativeResult.replace(/(薪)薪/g, '$1')
  superlativeResult = superlativeResult.replace(/褜$/g, '')
  return head + superlativeResult
}
