// natural/lib/natural/stemmers/porter_stemmer_uk.js
'use strict'

const Stemmer = require('./stemmer_uk')

const PorterStemmer = new Stemmer()
module.exports = PorterStemmer

function attemptReplacePatterns (token, patterns) {
  let replacement = null
  let i = 0; let isReplaced = false
  while ((i < patterns.length) && !isReplaced) {
    if (patterns[i][0].test(token)) {
      replacement = token.replace(patterns[i][0], patterns[i][1])
      isReplaced = true
    }
    i++
  }
  return replacement
}

function perfectiveGerund (token) {
  const result = attemptReplacePatterns(token, [
    [/[邪褟]胁(褕懈|褕懈褋褜)$/g, ''],
    [/(懈胁|懈胁褕懈|懈胁褕懈褋褜)$/g, '']
  ])
  return result
}

function adjectival (token) {
  let result = adjective(token)
  if (result != null) {
    const pariticipleResult = participle(result)
    result = pariticipleResult || result
  }
  return result
}

function adjective (token) {
  const result = attemptReplacePatterns(token, [
    [/(懈屑懈|褨泄|懈泄|邪|械|芯胁邪|芯胁械|褨胁|褦|褩泄|褦褦|械褦|褟|褨屑|械屑|懈屑|褨屑|懈褏|褨褏|芯褞|泄屑懈|褨屑懈|褍|褞|芯谐芯|芯屑褍|芯褩)$/g, '']
  ])
  return result
}

function participle (token) { // edit
  const result = attemptReplacePatterns(token, [
    [/(懈泄|芯谐芯|芯屑褍|懈屑|褨屑|邪|褨泄|褍|芯褞|褨泄|褨|懈褏|泄屑懈|懈褏)$/g, '']
  ])
  return result
}

function reflexive (token) {
  const result = attemptReplacePatterns(token, [
    [/(褋[褟褜懈])$/g, '']
  ])
  return result
}

function verb (token) {
  const result = attemptReplacePatterns(token, [
    [/(褋褜|褋褟|懈胁|邪褌褜|褟褌褜|褍|褞|邪胁|邪谢懈|褍褔懈|褟褔懈|胁褕懈|褕懈|械|屑械|邪褌懈|褟褌懈|褦)$/g, '']
  ])
  return result
}

function noun (token) {
  const result = attemptReplacePatterns(token, [
    [/(邪|械胁|芯胁|械|褟屑懈|邪屑懈|械懈|懈|械泄|芯泄|懈泄|泄|懈褟屑|褟屑|懈械屑|械屑|邪屑|芯屑|芯|褍|邪褏|懈褟褏|褟褏|褘|褜|懈褞|褜褞|褞|懈褟|褜褟|褟|褨|芯胁褨|褩|械褞|褦褞|芯褞|褦|械胁褨|械屑|褦屑|褨胁|褩胁|褞)$/g, '']
  ])
  return result
}

function superlative (token) {
  const result = attemptReplacePatterns(token, [
    [/(械泄褕|械泄褕械)$/g, '']
  ])
  return result
}

function derivational (token) {
  const result = attemptReplacePatterns(token, [
    [/[^邪械懈芯褍褞褟褨褩褦][邪械懈芯褍褞褟褨褩褦]+[^邪械懈芯褍褞褟褨褩褦]+[邪械懈芯褍褞褟褨褩褦].*(?<=芯)褋褌褜?$/g, '']
  ])
  return result
}

// perform full stemming algorithm on a single word
PorterStemmer.stem = function (token) {
  token = token.toLowerCase()
  const volwesRegexp = /^(.*?[邪械懈芯褍褞褟褨褩褦])(.*)$/g
  let RV = volwesRegexp.exec(token)
  if (!RV || RV.length < 3) {
    return token
  }
  const head = RV[1]
  RV = RV[2]
  volwesRegexp.lastIndex = 0
  const R2 = volwesRegexp.exec(RV)
  let result = perfectiveGerund(RV)
  if (result === null) {
    const resultReflexive = reflexive(RV) || RV
    result = adjectival(resultReflexive)
    if (result === null) {
      result = verb(resultReflexive)
      if (result === null) {
        result = noun(resultReflexive)
        if (result === null) {
          result = resultReflexive
        }
      }
    }
  }
  result = result.replace(/懈$/g, '')
  let derivationalResult = result
  if (R2 && R2[2]) {
    derivationalResult = derivational(R2[2])
    if (derivationalResult != null) {
      derivationalResult = derivational(result)
    } else {
      derivationalResult = result
    }
  }

  let superlativeResult = superlative(derivationalResult) || derivationalResult

  superlativeResult = superlativeResult.replace(/(薪)薪/g, '$1')
  superlativeResult = superlativeResult.replace(/褜$/g, '')
  return head + superlativeResult
}
